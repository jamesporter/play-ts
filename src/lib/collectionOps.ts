export function pairWise<T>(items: T[]): [T, T][] {
  if (items.length < 2) return [];
  const res: [T, T][] = [];
  for (let i = 0; i < items.length - 1; i++) {
    res.push([items[i], items[i + 1]]);
  }
  return res;
}

export function tripleWise<T>(items: T[]): [T, T, T][] {
  if (items.length < 3) return [];
  const res: [T, T, T][] = [];
  for (let i = 0; i < items.length - 2; i++) {
    res.push([items[i], items[i + 1], items[i + 2]]);
  }
  return res;
}
