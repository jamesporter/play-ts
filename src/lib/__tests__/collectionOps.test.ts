import { pairWise, tripleWise } from "../collectionOps";

describe("Pairwise", () => {
  it("should be able to do pairwise on simple array", () => {
    expect(pairWise([1, 2, 3, 4, 5])).toMatchSnapshot();
  });

  it("should give empty array when not long enough", () => {
    expect(pairWise([1])).toEqual([]);
  });
});

describe("Triplewise", () => {
  it("should be able to do triplewise on simple array", () => {
    expect(tripleWise([1, 2, 3, 4, 5])).toMatchSnapshot();
  });

  it("should give empty array when not long enough", () => {
    expect(tripleWise([1, 2])).toEqual([]);
  });
});
