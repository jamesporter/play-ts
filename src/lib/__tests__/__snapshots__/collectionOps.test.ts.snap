// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pairwise should be able to do pairwise on simple array 1`] = `
Array [
  Array [
    1,
    2,
  ],
  Array [
    2,
    3,
  ],
  Array [
    3,
    4,
  ],
  Array [
    4,
    5,
  ],
]
`;

exports[`Triplewise should be able to do triplewise on simple array 1`] = `
Array [
  Array [
    1,
    2,
    3,
  ],
  Array [
    2,
    3,
    4,
  ],
  Array [
    3,
    4,
    5,
  ],
]
`;
